Deprecate Whitelist for IBM-cloud-databases. Introduce allowlisting

omaraibrahim opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/IBM-Cloud/terraform-provider-ibm/blob/master/.github/CONTRIBUTING.md --->
In the api version 5 whitelist was deprecated and allowlist was introduced. In our strive for feature parity, in the following PR I deprecated whitelist and introduced allowlist.

This involved:
- Deprecating whitelist in our data_source_ibm_database and introducing allowlist
- Deprecating whitelist in our database resource and introducing allowlist
- Updating tests

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Relates OR Closes #0000

Output from acceptance testing:

<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->
```
$ make testacc TEST=./ibm/service/database TESTARGS='-run=TestAccIBMDatabaseDataSource_basic'
--- PASS: TestAccIBMDatabaseDataSource_basic (597.39s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/database	598.703s
...
```

```
$ make testacc TEST=./ibm/service/database TESTARGS='-run=TestAccIBMEDBDatabaseInstanceBasic'
--- PASS: TestAccIBMEDBDatabaseInstanceBasic (1836.72s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/database	1837.595s
```

```
$ make testacc TEST=./ibm/service/database TESTARGS='-run=TestAccIBMDatabaseInstancePostgresBasic'
--- PASS: TestAccIBMDatabaseInstancePostgresBasic (1493.88s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/database	1494.758s
```

```
$ make testacc TEST=./ibm/service/database TESTARGS='-run=TestAccIBMDatabaseInstancePostgresAllowlistMigration'
--- PASS: TestAccIBMDatabaseInstancePostgresAllowlistMigration (665.06s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/database        668.257s
```

hkantare commented on ibm/service/database/data_source_ibm_database.go:
```diff
@@ -700,11 +720,26 @@ func dataSourceIBMDatabaseInstanceRead(d *schema.ResourceData, meta interface{})
-	whitelist, err := icdClient.Whitelists().GetWhitelist(icdId)
-	if err != nil {
-		return fmt.Errorf("[ERROR] Error getting database whitelist: %s", err)
+	if _, ok := d.GetOk("whitelist"); ok {
```
whitelist and allowlist are computed variables you can't do d.GetOk...

hkantare replied:
can we use cloud v5 API and flatten the result back to both whitelist and allowlist

hkantare commented on ibm/service/database/resource_ibm_database.go:
```diff
@@ -459,6 +459,29 @@ func ResourceIBMDatabaseInstance() *schema.Resource {
+				Deprecated:    "Whitelist is deprecated please use allowlist",
+				ConflictsWith: []string{"allowlist"},
+			},
+			"allowlist": {
```
Can't we set the v5 API response of allowlist to whitelist (because the schema looks same for both) it just flatten and setting the same

omaraibrahim replied:
Hi @hkantare!

They're two different apis endpoints, and despite them being very similar, I think it's safer to keep them separate. Whitelist is depreciated and will be removed soon as well. Hence, I think a separation of concerns is better. Here are the api docs for both. https://github.com/IBM-Cloud/terraform-provider-ibm/pull/3852#discussion_r902894675

Thanks!
Omar

hkantare replied:
Yes ...v4/v5 API are two different endpoint with same model structure and functionality ...cant we just handle in the backend (i mean resource/datasource) to migrate to V5 AP's without any change to schema.

omaraibrahim replied:
@hkantare so the responses are the same yes :D But the supporting functionality is not. Meaning, the autogenerated allowlist functions utilize allowlist structs the use pointers. For example, https://github.com/IBM/cloud-databases-go-sdk/blob/de91b81e136fd4c53c0000e87932aaf150658a33/clouddatabasesv5/cloud_databases_v5.go#L1848

In contrast the whitelist functions utilizes a whitelist struct that does not use pointers. An example of the functions that consume structs that are shaped differently can  be found here: https://github.com/IBM-Cloud/bluemix-go/blob/d538cb4fd9bece50f7b012496df63b3f62d58662/api/icd/icdv4/whitelist.go#L38

So despite achieving the same result, the code is structured very differently and the structs that allowlist relies on are different than whitelist. For example, `AllowlistEntry` is written as follows: https://github.com/IBM/cloud-databases-go-sdk/blob/de91b81e136fd4c53c0000e87932aaf150658a33/clouddatabasesv5/cloud_databases_v5.go#L2046
While `WhitelistEntry` is written as follows https://github.com/IBM-Cloud/bluemix-go/blob/d538cb4fd9bece50f7b012496df63b3f62d58662/api/icd/icdv4/whitelist.go#L13

My concern of just setting the v5 response of allowlist to whitelist is we'll have to do lots of downstream changes that will break things or cause unexpected bugs.

omaraibrahim replied:
Oh I also test the instances using import and it works :)

hkantare commented on ibm/service/database/resource_ibm_database.go:
```diff
@@ -1710,11 +1763,25 @@ func resourceIBMDatabaseInstanceRead(context context.Context, d *schema.Resource
-	whitelist, err := icdClient.Whitelists().GetWhitelist(icdId)
-	if err != nil {
-		return diag.FromErr(fmt.Errorf("[ERROR] Error getting database whitelist: %s", err))
+	if _, ok := d.GetOk("whitelist"); ok {
```
In Resource Read also don't use d.Get operations because the import functionality doesn't work

hkantare replied:
Thanks for explanation
getOK does work in resource read  but it fails during the import scenario. Let say you have created  a database outside Terraform and want to import and  manage in future using Terraform. We need to import the resource. (https://learn.hashicorp.com/tutorials/terraform/state-import?in=terraform/state&utm_source=WEBSITE&utm_medium=WEB_IO&utm_offer=ARTICLE_PAGE&utm_content=DOCS)
For import we first create a empty resource block
resource "ibm_database" "mydb" {
}
Run terraform import ibm_database.mydb <id>
Import also calls the same resource read since the block is empty d.GetOk fails for both whitelist and allowlist

omaraibrahim replied:
@hkantare thank you very very much for the explanation :D I've gone ahead and update the  code so that if whitelist isn't there we always use allowlist. I follow the terraform docs here: https://www.terraform.io/plugin/sdkv2/best-practices/deprecations#renaming-an-optional-attribute

So if `d.getOK("whitelist")` fails we'll always use allowlist. lmk what you think :D

hkantare commented:
@omaraibrahim Can you add a migration testcase from whitelist to allowlist

omaraibrahim commented:
@hkantare I've gone ahead and added migration tests. I've also modified the code to address some concerns and ensure a seemless migration to `allowlist` - while retaining full functionality for `whitelist`.
